README.md
# @elschnagoo/fritzapi
> **Forked** rework in **Typescript** and more modern code style


[![NPM Version](https://img.shields.io/npm/v/@elschnagoo/fritzapi.svg)](https://www.npmjs.com/package/@elschnagoo/fritzapi)
[![NPM Downloads](https://img.shields.io/npm/dt/@elschnagoo/fritzapi.svg)](https://www.npmjs.com/package/@elschnagoo/fritzapi)


## Install
    npm i @elschnagoo/fritzapi

## Links

- [Docs](https://elschnagoo.github.io/fritzapi/)
- [TestReport](https://elschnagoo.github.io/fritzapi/test)
- [Coverage](https://elschnagoo.github.io/fritzapi/coverage)


## New example in Ts:
```ts
import Fritz from '@elschnagoo/fritzapi'

const f = new Fritz("user", "password", "http://192.168.178.1");

(async ()=>{
    const sid = await f.getSID();
    console.log(sid)

    const list = await f.getSwitchList();
    console.log(list)
})()

```
# Original README.MD (Not maintained anymore)

[![NPM Version](https://img.shields.io/npm/v/fritzapi.svg)](https://www.npmjs.com/package/fritzapi)
[![NPM Downloads](https://img.shields.io/npm/dt/fritzapi.svg)](https://www.npmjs.com/package/fritzapi)
[![Build status](https://travis-ci.org/andig/fritzapi.svg?branch=master)](https://travis-ci.org/andig/fritzapi)


Home automation node API for Fritz!Box, Fritz!DECT and FRITZ!Powerline devices.

[homebridge-fritz](https://github.com/andig/homebridge-fritz) and [pimatic-fritz](https://github.com/andig/pimatic-fritz) are built on fritzapi.

## Functions

### General functions

- Get the Fritz!OS version `getOSVersion`
- Get the session ID `getSessionID`
- Get device list as XML `getDeviceListInfos` >Fritz!OS 6.10
- Get device list `getDeviceList` >Fritz!OS 6.10
- Get device list with filter criteria applied `getDeviceListFiltered` >Fritz!OS 6.10
- Get device `getDevice` >Fritz!OS 6.10
- Get temperature `getTemperature` - polyfill
- Get presence `getPresence` - polyfill
- Get template list as XML `getTemplateListInfos` >Fritz!OS 7.0
- Get template list `getTemplateList` >Fritz!OS 7.0
- Get template list `applyTemplate` >Fritz!OS 7.0

**Note**

`getTemperature` works for DECT repeaters but is not available on the FRITZ!Powerline 546E WLAN set and will always return `NaN`.

While `getTemperature` works for outlets, it is not available for (outlet) groups that can be created through the Fritz!Box user interface.

`getDeviceListInfos` was named `getDeviceListInfo` in earlier versions. For consistency with the official Fritz!Box API the name has been changed. The `getDeviceListInfo` name is deprecated and will be removed in a future release.
In general, use of `getDeviceListInfos` is discouraged as the equivalent `getDeviceList` function which returns an object interface instead of XML is easier to use.


### Fritz!DECT 200 and 210 outlet functions (includes FRITZ!Powerline 546E)

- Get list `getSwitchList`
- Get state `getSwitchState`
- Set on `setSwitchOn`
- Set off `setSwitchOff`
- Toggle `setSwitchToggle`
- Get power `getSwitchPower`
- Get energy `getSwitchEnergy`
- Get presence status `getSwitchPresence`
- Get name `getSwitchName`
- Get basic device stats as XML `getBasicDeviceStats` >Fritz!OS 7.0

For controlling AVM Fritz!DECT 200 devices the actuator identification number (AIN) is needed. The AIN can be obtained using `getSwitchList` which returns a list of AINs or the more general `getDeviceList` function which returns a verbose device list structure as JSON.


### Fritz!DECT 100 functions

The Fritz!DECT 100 DECT repeater AIN does only appear in the `getDeviceList` output. It supports retrieving the repeater's temperature.


### Fritz!DECT 300, 301 and CometDECT thermostat functions

Thermostat functions are only available as of Fritz!OS 6.36

- Get list `getThermostatList` - polyfill
- Set target temperature `setTempTarget`, supports 'ON'/'OFF' to enable/disable thermostat
- Get target temperature `getTempTarget`
- Get comfort temperature `getTempComfort`
- Get night temperature `getTempNight`
- Get battery charge `getBatteryCharge` (uses UI scraping, may be unstable)
- Get window open `getWindowOpen` (uses UI scraping, may be unstable)

### Fritz!DECT 500 and Telekom bulb functions

Bulb functions are only available as of Fritz!OS 7.20

- Get List of all bulbs `getBulbList`
- Get List of all bulbs supporting colors `getColorBulbList`
- Switch bulb on/off or toggle it `setSimpleOnOff`
- Dimm a bulb `setLevel`, 0 - 255
- Dimm a bulb `setLevelPercentage`, 0 - 100
- Set the color and saturation `setColor`, see note 1
- Set the color temperature `setColorTemperature`, see note 2

**Note 1**
The color api calls of the FritzBox accept only a predefined set of values.
To get easy access to the predefined values fritzapi accepts the following values as the color:
`red, orange, yellow, lime, green, turquoise, cyan, lightblue, blue, purple, magenta, pink`
The saturation can be `0, 1 or 2` to select one of the predefined saturation values for the color.

**Note 2**
The color temperature api calls of the FritzBox accept only a predefined set of values.
Valid values are `2700, 3000, 3400,3800, 4200, 4700, 5300, 5900 and 6500`.
Other values are adjusted by fritzapi to one of the above values.

### WLAN functions

- Get the guest wlan settings `getGuestWlan`
- Set the guest wlan `setGuestWlan`

**Note**

`getGuestWlan` returns a structure containing all wifi settings found in the Fritz!Box UI. The `setGuestWlan` function accepts either a settings structure such as this or a single boolean value.

All functions have been tested on Fritz!OS 6.20/6.36/6.51 using the Fritz!Box 7390 and on Fritz!OS 6.50 on Fritz!Box 6490. The WLAN functions may be less stable.


## Installation

```bash
npm install fritzapi
```


## Usage

### Object-oriented interface

The object-oriented interface is the recommended way of using fritzapi.

Get the list of switch AINs using a customer Fritz!Box address:
```js
var Fritz = require('fritzapi').Fritz;

var f = new Fritz("user", "password", "http://192.168.178.1");

f.getSwitchList().then(function(ains){
  console.log(f.getSID());
  console.log(ains);
});
```

### Functional interface

Get the session ID using default Fritz!Box address (http://fritz.box):
```js
var fritz = require('fritzapi');

fritz.getSessionID("user", "password").then(function(sid) {
    console.log(sid);
});
```

Get the list of switch AINs using a custom Fritz!Box address with self-signed certificate:
```js
fritz.getSessionID("user", "password", {
  url: "192.168.178.1",
  strictSSL: false         // workaround DEPTH_ZERO_SELF_SIGNED_CERT SSL error
}).then(function(sid) {
  console.log(sid);

  // note that the options/url need be carries through every single api call
  fritz.getSwitchList(sid, options).then(function(ains){
    console.log(ains);
  });
});
```

**Note** The functional interface may be deprecated in a future version of the library.


## Device details

Below is the output of `getDeviceList` for reference.

The list was produced for various Fritz devices I've had around. It might have changed in the meantime depending on device firmware or Fritz HTTP API version.
These definitions remain cached by the Fritz!Box even if the device is no longer connected. The device presence is indicated by the `present` attribute.

### Powerline

    { identifier: '34:31:C4:DB:F6:C7',
      id: '20000',
      functionbitmask: '640',
      fwversion: '06.20',
      manufacturer: 'AVM',
      productname: 'FRITZ!Powerline 546E',
      present: '0',
      name: 'FRITZ!Powerline 546E',
      switch: { state: '', mode: '', lock: '' },
      powermeter: { power: '', energy: '' } }

### Outlets

    { identifier: '087610103568',
      id: '17',
      functionbitmask: '35712',
      fwversion: '04.16',
      manufacturer: 'AVM',
      productname: 'FRITZ!DECT 200',
      present: '1',
      name: 'FRITZ!DECT 200 #1',
      switch: { state: '', mode: '', lock: '', devicelock : '' },
      simpleonoff { state: '' },
      powermeter: { voltage: '', power: '', energy: '' },
      temperature: { celsius: '', offset: '' }
      }

    { identifier: '116570031825',
      id: '18',
      functionbitmask: '640',
      fwversion: '03.67',
      manufacturer: 'AVM',
      productname: 'FRITZ!DECT 210',
      present: '0',
      name: 'FRITZ!DECT 210 #3',
      switch: { state: '', mode: '', lock: '' },
      powermeter: { power: '', energy: '' } }

### DECT Repeater

    { identifier: '116570002527',
      id: '20',
      functionbitmask: '1024',
      fwversion: '03.64',
      manufacturer: 'AVM',
      productname: 'FRITZ!DECT Repeater 100',
      present: '0',
      name: 'FRITZ!DECT Rep 100 #5' }

### Thermostats

    { identifier: '109710195784',
      id: '17',
      functionbitmask: '320',
      fwversion: '03.54',
      manufacturer: 'AVM',
      productname: 'Comet DECT',
      present: '1',
      txbusy: '0'
      name: 'Comet DECT',
      battery: '90'
      batterylow: '0'
      temperature: { celsius: '', offset: '' },
      hkr: { tist: '49', tsoll: '253', absenk: '32', komfort: '42', lock: '1', devicelock: '1', errorcode: '0',
             windowopenactiv: '0', windowopenactiveendtime: '0', boostactive: '0', boostactiveendtime: '0',
             battery: '90', batterylow: '0', summeractive: '1', holidayactive: '0',
             nextchange: { endperiod: '1598907600', tchange: '32' }} }

### HANFUN

As of Fritz!OS 7 the HANFUN devices have their own bitmask `1`:

    { identifier: '119340326786',
      id: '406',
      functionbitmask: '1',
      fwversion: '00.00',
      manufacturer: '0x0feb',
      productname: 'HAN-FUN',
      present: '1',
      name: 'HAN-FUN #1' }

HANFUN functions are accessible as "HANFUN unit" devices. Bitmask consists of HANFUN unit (bit 13) plus actual function (in this case ALARM, bit 3):

    { identifier: '119340326786-1',
      id: '2000',
      functionbitmask: '8208',
      fwversion: '0.0',
      manufacturer: '0x0feb',
      productname: 'HAN-FUN',
      present: '1',
      name: 'no name',
      etsiunitinfo: { etsideviceid: '406', unittype: '514', interfaces: '256' },
      alert: { state: '1' } }

or a color bulb:

    { identifier: '127010028190-1',
      id: '2000',
      functionbitmask: '237572',
      fwversion: '0.0',
      manufacturer: '0x319d',
      productname: 'HAN-FUN',
      present: '0',
      txbusy: '0'
      name: 'Bulb 1',
      simpleonoff { state: '1' },
      levelcontrol: { level: '255', levelpercentage: '100' },
      colorcontrol: { supported_modes: '5', current_mode: '1', hue: '35', saturation: '214', temperature: '' },
      etsiunitinfo: { etsideviceid: '406', unittype: '278', interfaces: '512,514,513' } }

## AHA HTTP Interface Documentation

http://avm.de/fileadmin/user_upload/Global/Service/Schnittstellen/AHA-HTTP-Interface.pdf

## Acknowledgements

Thanks to:

* AVM for providing free test devices and the AHA-HTTP interface document
* EUROtronic Technology GmbH for providing free CometDECT thermostat sample
* nischelwitzer for the basic js implementation (https://github.com/nischelwitzer/smartfritz)
* steffen.timm for the basic communication function
* thk4711 for the FRITZ!DECT 200 codes


CHANGELOG.md
## 0.13.7

- Upgrade dependencies

## 0.13.3

- Extending AVM Device Interfaces
- Device validation function `isAVMSwitch`,`isAVMBulb`

## 0.13.0

- Add color type
- Add more unit tests
- Add full OSInfoInterface
- Add more test env variables
- Add `eslint` airbnb config
- Add `typedoc` for documentation
- Replace `mocha` & `istanbul` with `jest`
- Replace `request` with `axios`
- Replace `xml2json-light` with `xml2json`
- Remove `bluebird` use native Promise
- Remove `cheerio`
- Fix post form format
- Rewrite legacy js in `typescript`

## 0.12.0

- enable blind controll

## 0.10.3

- expose temperature conversion helpers

## 0.10.2

- bugfix `getThermostatList` not relaying `options`

## 0.10.1

- added `getWindowOpen` polyfill

## 0.10.0

- renamed `getDeviceListInfo` to `getDeviceListInfos`
- changed `getDeviceList` and `getDevice` to always return AINs without spaces. These AINs can directly be passed into the native Fritz!Box api functions.
- added `getDeviceListFiltered` which takes a `filter` object to apply to the devices. Used internally for filtering by `functionbitmask`.

## 0.9.6

- bugfixed OO bitfunctions in object

- moved jslint to pretest
- added first testcase (mocha+chai based)
- added testdir ./test
- npm test testcase look up

- travis workflow: before_install: jslint, before_script: npm test, after_sucess: istanbul coverage
- added istanbul cover (npm run coverage)

## 0.9.0

- **Breaking change**: SSL certificate check is now strict. This will break self-signed certificates that are typically used by AVM Fritz!Box. To connect via HTTPS to such a Fritz!Box add the `strictSSL: false` option:

	fritz.getSessionID(username, password, {
		url: "https://...",
		strictSSL: false
	}).then(function(sid) {
		// ...
	});

- changed `getSwitchPower` and `getBatteryCharge` to return `null` instead of 0 or empty if device is not connected

## 0.8.0

- added object interface


src/Fritz.ts
/**
 * smartFritz - Fritz goes smartHome
 *
 * AVM SmartHome nodeJS Control - for AVM Fritz!Box and Dect200 Devices
 *
 * @author Andreas Goetz <cpuidle@gmx.de>
 *
 * Forked from: https://github.com/nischelwitzer/smartfritz
 * nischi - first version: July 2014
 *
 * based on: Node2Fritz by steffen.timm on 05.12.13 for Fritz!OS > 05.50
 * and  thk4711 (code https://github.com/pimatic/pimatic/issues/38)
 *
 * AVM Documentation is at https://avm.de/service/schnittstellen/
 */

/*
 * Object-oriented API
 */

import axios from 'axios';
import crypto from 'crypto';
import parser from 'xml2json';
import * as qs from 'qs';
import {
  api2temp,
  AVMDevice,
  color2apihue,
  ColorName,
  colortemp2api,
  defaultUrl,
  FUNCTION_BUTTON,
  FUNCTION_COLORCONTROL,
  FUNCTION_LIGHT,
  FUNCTION_THERMOSTAT,
  IBase,
  IExt,
  IState,
  ITemp,
  level2api,
  OSInfo,
  ReqOption,
  satindex2apisat,
  state2api,
  temp2api,
  time2api,
} from './utils';

export default class Fritz {
  protected sid: string | null;

  protected username: string;

  protected password: string;

  protected url = defaultUrl;

  protected deviceList: AVMDevice[] | null;

  /**
   * Default constructor
   * @param username
   * @param password
   * @param url if not set default is 'http://fritz.box'
   */
  constructor(username: string, password: string, url?: string) {
    this.sid = null;
    this.username = username;
    this.password = password;
    this.deviceList = null;
    if (url) {
      this.url = url;
    }
  }

  /**
   * get SID
   */
  async getSID(): Promise<string> {
    if (this.sid === null) {
      this.sid = await this.getSessionID(this.username, this.password);
    }
    if (this.sid === null) {
      throw new Error('Bad SID');
    }
    return this.sid;
  }

  /**
   * Request helper
   * @param request
   * @param options
   * @private
   */
  private static async httpRequest<T>(
    request: string,
    options?: ReqOption,
  ): Promise<T> {
    try {
      let res;
      if (options?.method === 'POST') {
        if (options.form) {
          const form = qs.stringify(options.form);

          res = await axios.post<any>(request, form, {
            headers: {
              'Content-Type': 'application/x-www-form-urlencoded',
            },
          });
        } else {
          throw new Error('no form data');
        }
      } else {
        res = await axios.get<any>(request);
      }
      return res.data;
    } catch (e) {
      console.error(e);
      throw new Error('httpRequest Failed');
    }
  }

  /**
   * Execute Fritz API command for device specified by AIN
   */
  async executeCommand<T>(
    sid: boolean,
    command: string | null,
    ain: string | null,
    path?: string,
    ...param: string[]
  ) {
    const ePath = [];
    const eParam = [];
    ePath.push(this.url);
    if (path) {
      ePath.push(path);
    } else {
      ePath.push('/webservices/homeautoswitch.lua');
      eParam.push('0=0');
    }
    param.forEach((el) => {
      eParam.push(el);
    });

    if (sid) eParam.push(`sid=${await this.getSID()}`);
    if (command) eParam.push(`switchcmd=${command}`);
    if (ain) eParam.push(`ain=${ain}`);

    if (eParam.length > 0) {
      ePath.push('?');
      ePath.push(eParam.join('&'));
    }
    console.log(ePath.join(''));
    return Fritz.httpRequest<T>(ePath.join(''));
  }

  /**
   * get session id
   * @param username
   * @param password
   * @private
   */
  private async getSessionID(
    username: string,
    password: string,
  ): Promise<string> {
    let body = await this.executeCommand<any>(
      false,
      null,
      null,
      '/login_sid.lua',
    );

    const challenge = body.match('<Challenge>(.*?)</Challenge>')[1];
    const challengeResponse = `${challenge}-${crypto
      .createHash('md5')
      .update(Buffer.from(`${challenge}-${password}`, 'utf16le'))
      .digest('hex')}`;
    const url = `/login_sid.lua?username=${username}&response=${challengeResponse}`;

    body = await this.executeCommand(false, null, null, url);

    const sessionID = body.match('<SID>(.*?)</SID>')[1];
    return sessionID;
  }

  /**
   * check if session id is OK
   */
  async checkSession(): Promise<boolean> {
    const body = await this.executeCommand<string>(
      true,
      null,
      null,
      '/login_sid.lua',
    );
    const sessionID = body.match('<SID>(.*?)</SID>');
    return !!sessionID && sessionID[1] !== '0000000000000000';
  }

  /**
   *  get OS version
   */
  async getOSVersion() {
    const req: ReqOption = {
      method: 'POST',
      form: {
        sid: await this.getSID(),
        xhr: 1,
        page: 'overview',
      },
    };
    const json = await Fritz.httpRequest<OSInfo>(`${this.url}/data.lua`, req);
    const osVersion =
      json.data && json.data.fritzos && json.data.fritzos.nspver
        ? json.data.fritzos.nspver
        : null;
    return osVersion;
  }

  /**
   * get full FritzBox config
   */
  async getDataSheet(): Promise<OSInfo> {
    const req: ReqOption = {
      method: 'POST',
      form: {
        sid: await this.getSID(),
        xhr: 1,
        page: 'overview',
      },
    };
    return Fritz.httpRequest<OSInfo>(`${this.url}/data.lua`, req);
  }

  /**
   * get template information (XML)
   */
  async getTemplateListInfos(): Promise<string> {
    return this.executeCommand<string>(true, 'gettemplatelistinfos', null);
  }

  /**
   *  get template information (json)
   */
  getTemplateList: IBase<any[]> = async () => {
    const templateinfo = await this.getTemplateListInfos();

    const templates = JSON.parse(parser.toJson(templateinfo));

    // extract templates as array
    let out: any[] = [];
    out = out.concat((templates.templatelist || {}).template || []);
    out = out.map((template) => {
      return template;
    });
    return out;
  };

  /**
   *  apply template
   * @param ain
   */
  async applyTemplate(ain: string) {
    return this.executeCommand(true, 'applytemplate', ain);
  }

  /**
   * get basic device info (XML)
   * @param ain
   */
  async getBasicDeviceStats(ain: string) {
    return this.executeCommand(true, 'getbasicdevicestats', ain);
  }

  /**
   *  get detailed device information (XML)
   */

  async getDeviceListInfos() {
    return this.executeCommand<string>(true, 'getdevicelistinfos', null);
  }

  /**
   * get device list
   */
  async getDeviceList(): Promise<AVMDevice[]> {
    const devicelistinfo = await this.getDeviceListInfos();
    const devicesCore = JSON.parse(parser.toJson(devicelistinfo));
    // extract devices as array
    let devices: AVMDevice[] = [];
    devices = devices.concat((devicesCore.devicelist || {}).device || []);
    devices = devices.map((dev: any) => {
      const newElement = dev;
      newElement.identifier = newElement.identifier.replace(/\s/g, '');
      return newElement;
    });
    return devices;
  }

  /**
   * get device list by filter criteria
   * @param filter
   */
  async getDeviceListFiltered(filter: any): Promise<AVMDevice[]> {
    if (!this.deviceList) {
      this.deviceList = await this.getDeviceList();
    }

    return this.deviceList.filter((device: any) => {
      return Object.keys(filter).every(function (key) {
        /* jshint laxbreak:true */
        return key === 'functionbitmask'
          ? device.functionbitmask & filter[key]
          : device[key] === filter[key];
      });
    });
  }

  /**
   *  get single device
   * @param ain
   */
  async getDevice(ain: string): Promise<AVMDevice | null> {
    const devices = await this.getDeviceListFiltered({
      identifier: ain,
    });
    return devices.length ? devices[0] : null;
  }

  /**
   * get temperature- both switches and thermostats are supported, but not powerline modules
   * @param ain
   */
  async getTemperature(ain: string): Promise<number> {
    const body = await this.executeCommand<string>(true, 'gettemperature', ain);
    return parseFloat(body) / 10; // °C
  }

  /**
   * get presence from deviceListInfo
   * @param ain
   */
  async getPresence(ain: string): Promise<boolean> {
    return !!(await this.getDevice(ain))?.presence;
  }

  /*
   * Switches
   */

  /**
   * get switch list
   */
  async getSwitchList(): Promise<any[]> {
    let res = await this.executeCommand<string>(true, 'getswitchlist', null);
    res = res.replace('\n', '');
    return res === '' ? [] : res.split(',');
  }

  /**
   * get switch state
   * @param ain
   */
  getSwitchState: IExt<boolean> = async (ain) => {
    const body = await this.executeCommand<string>(true, 'getswitchstate', ain);
    return /^1/.test(body); // true if on
  };

  /**
   * turn an outlet on. returns the state the outlet was set to
   * @param ain
   */
  setSwitchOn: IExt<boolean> = async (ain) => {
    const body = await this.executeCommand<string>(true, 'setswitchon', ain);
    return /^1/.test(body); // true if on
  };

  /**
   * turn an outlet off. returns the state the outlet was set to
   * @param ain
   */
  setSwitchOff: IExt<boolean> = async (ain) => {
    const body = await this.executeCommand<string>(true, 'setswitchoff', ain);
    return /^1/.test(body); // false if off
  };

  /**
   * toggle an outlet. returns the state the outlet was set to
   * @param ain
   */
  setSwitchToggle: IExt<boolean> = async (ain) => {
    const body = await this.executeCommand<string>(
      true,
      'setswitchtoggle',
      ain,
    );
    return /^1/.test(body); // false if off
  };

  /**
   * get the total enery consumption. returns the value in Wh
   * @param ain
   */
  getSwitchEnergy: IExt<number> = async (ain) => {
    const body = await this.executeCommand<string>(
      true,
      'getswitchenergy',
      ain,
    );
    return parseFloat(body); // Wh
  };

  /**
   * get the current enery consumption of an outlet. returns the value in mW
   * @param ain
   */
  getSwitchPower: IExt<number | null> = async (ain) => {
    const body = await this.executeCommand<string>(true, 'getswitchpower', ain);
    const power = parseFloat(body);
    return Number.isNaN(power) ? null : power / 1000; // W
  };

  /**
   * get the outet presence status
   * @param ain
   */
  getSwitchPresence: IExt<boolean> = async (ain) => {
    const body = await this.executeCommand<string>(
      true,
      'getswitchpresent',
      ain,
    );
    return /^1/.test(body); // true if present
  };

  /**
   * get switch name
   * @param ain
   */
  getSwitchName: IExt<string> = async (ain) => {
    const body = await this.executeCommand<string>(true, 'getswitchname', ain);
    return body.trim();
  };

  /*
   * Thermostats
   */

  /**
   * get the thermostat list
   */
  getThermostatList: IBase<any> = async () => {
    const devices = await this.getDeviceListFiltered({
      functionbitmask: FUNCTION_THERMOSTAT,
    });
    return devices.map((device) => {
      return device.identifier;
    });
  };

  /**
   * set target temperature (Solltemperatur)
   * @param ain
   * @param temp
   */
  async setTempTarget(ain: string, temp: number): Promise<number> {
    await this.executeCommand(true, `sethkrtsoll&param=${temp2api(temp)}`, ain);
    return temp;
  }

  /**
   * get target temperature (Solltemperatur)
   * @param ain
   */
  getTempTarget: IExt<ITemp> = async (ain) => {
    const body = await this.executeCommand<string>(true, 'gethkrtsoll', ain);
    return api2temp(body);
  };

  /**
   * get night temperature (Absenktemperatur)
   * @param ain
   */
  getTempNight: IExt<ITemp> = async (ain) => {
    const body = await this.executeCommand<string>(true, 'gethkrabsenk', ain);
    return api2temp(body);
  };

  /**
   * get comfort temperature (Komforttemperatur)
   * @param ain
   */
  getTempComfort: IExt<ITemp> = async (ain) => {
    const body = await this.executeCommand<string>(true, 'gethkrkomfort', ain);
    return api2temp(body);
  };

  /**
   * ------------------------------------------------
   */
  /**
   * Not yet tested - deactivated for now
   */
  //
  /**
   * activate boost with end time or deactivate boost
   * @param ain
   * @param endtime
   */
  async setHkrBoost(ain: string, endtime: number) {
    await this.executeCommand(
      true,
      `sethkrboost&endtimestamp=${time2api(endtime)}`,
      ain,
    );
    return endtime;
  }

  /**
   * activate window open  with end time or deactivate boost
   * @param ain
   * @param endtime
   */
  async setHkrWindowOpen(ain: string, endtime: number) {
    await this.executeCommand(
      true,
      `sethkrwindowopen&endtimestamp=${time2api(endtime)}`,
      ain,
    );
  }

  /**
   * activate window open  with end time or deactivate boost
   * @param deviceId
   * @param offset
   */
  async setHkrOffset(deviceId: string, offset: number) {
    const path = '/net/home_auto_hkr_edit.lua';

    const req: ReqOption = {
      method: 'POST',
      form: {
        sid: await this.getSID(),
        xhr: 1,
        lang: 'de',
        no_sidrenew: '',
        ule_device_name: '',
        useajax: 1,
        WindowOpenTimer: '',
        WindowOpenTrigger: '',
        tempsensor: '',
        Roomtemp: '',
        ExtTempsensorID: '',
        Offset: offset,
        back_to_page: 'sh_dev',
        view: '',
        graphState: 1,
        apply: '',
        device: deviceId,
        oldpage: '/net/home_auto_hkr_edit.lua',
      },
    };

    await Fritz.httpRequest(path, req);
    return offset;
  }

  /**
   * ------------------------------------------------
   */

  /*
   * AVM Buttons Fritz!DECT 400 and Fritz!DECT 440
   * Querying a button isn't really useful because they don't have a state to query,
   * there is just a timestamp of the last short and long button press.
   * The only useful information is the battery status returned in the 'battery' and
   * 'batterylow' property.
   * The Fritz!DECT 440 should have an additional 'temperature' property
   */

  /**
   * get a list of all button devices
   */
  getButtonList: IBase<string[]> = async () => {
    const devices = await this.getDeviceListFiltered({
      functionbitmask: FUNCTION_BUTTON,
    });
    return devices.map((device) => {
      return device.identifier;
    });
  };

  /*
   * Light bulbs (HAN-FUN)
   */

  /**
   * get a list of all bulbs
   */
  getBulbList: IBase<string[]> = async () => {
    const devices = await this.getDeviceListFiltered({
      functionbitmask: FUNCTION_LIGHT,
    });
    return devices.map((device) => {
      return device.identifier;
    });
  };

  /**
   * get a list of bulbs which support colors
   */
  getColorBulbList: IBase<string[]> = async () => {
    const devices = await this.getDeviceListFiltered({
      functionbitmask: FUNCTION_LIGHT || FUNCTION_COLORCONTROL,
    });
    return devices.map((device) => {
      return device.identifier;
    });
  };

  /**
   * switch the device on, of or toggle its current state
   * @param ain
   * @param state
   */
  async setSimpleOnOff(ain: string, state: IState) {
    /**
     * ain = ain.replace('-1','');
     */
    await this.executeCommand(
      true,
      `setsimpleonoff&onoff=${state2api(state)}`,
      ain,
    );
    return state;
  }

  /**
   * Dimm the device, allowed values are 0 - 255
   * @param ain
   * @param level
   */
  async setLevel(ain: string, level: number) {
    await this.executeCommand(
      true,
      `setlevel&level=${level2api(level, false)}`,
      ain,
    );
    return level;
  }

  /**
   * Dimm the device, allowed values are 0 - 100
   * @param ain
   * @param levelInPercent
   */
  async setLevelPercentage(ain: string, levelInPercent: number) {
    return this.executeCommand(
      true,
      `setlevelpercentage&level=${level2api(levelInPercent, true)}`,
      ain,
    ).then(function () {
      /**
       * api does not return a value
       */
      return levelInPercent;
    });
  }

  /**
   *  Set the color and saturation of a color bulb
   * Valid color values are:
   * red, orange, yellow, lime, green, turquoise, cyan,
   * lightblue, blue, purple, magenta and pink
   * Valid satindex values are: 0, 1 or 2
   * @param ain
   * @param color
   * @param satindex
   * @param duration
   */
  async setColor(
    ain: string,
    color: ColorName,
    satindex: number,
    duration: number,
  ) {
    await this.executeCommand(
      true,
      `setcolor&hue=${color2apihue(color)}&saturation=${satindex2apisat(
        color,
        satindex,
      )}&duration=${duration}`,
      ain,
    );
    return color;
  }

  /**
   * Set the color temperature of a bulb.
   * @param ain
   * @param temperature Valid values are 2700, 3000, 3400,3800, 4200, 4700, 5300, 5900 and 6500.
   * @param duration
   */
  async setColorTemperature(
    ain: string,
    temperature: number,
    duration: number,
  ) {
    const temp = colortemp2api(temperature);
    await this.executeCommand(
      true,
      `setcolortemperature&temperature=${temp}&duration=${duration}`,
      ain,
    );
    return temp;
  }

  // get the color defaults
  // This is mostly useless because they are no defaults which can be changed but
  // fixed values. Only combinations returned by this api call are accepted by
  // setcolor and setcolortemperature.
  // module.exports.getColorDefaults = function(sid, ain, options)
  // {
  //     return executeCommand(true, 'getcolordefaults', ain, options).then(function(body) {
  //         return body;
  //     });
  // };

  // ------------------------------------------------
  // Not yet tested - deactivated for now
  // I don't know about any blind control unit with HANFUN support, but this API call makes
  // it plausible that AVM or a partner has somthing like that in the pipeline.
  //
  async setBlind(ain: string, blindState: string | number) {
    // „open“, „close“ or „stop“
    await this.executeCommand(true, `setblind&target=${blindState}`, ain);
    return blindState;
  }

  /**
   *  get battery charge
   * Attention: this function queries the whole device list to get the value for one device.
   * If multiple device will be queried for the battery status, a better approach would be to
   * get the device list once and then filter out the devices of interest.
   * @param ain
   */
  getBatteryCharge: IExt<unknown | null> = async (ain) => {
    const device = await this.getDevice(ain);
    return device?.battery;
  };

  /**
   * Get the window open flag of a thermostat
   * Attention: this function queries the whole device list to get the value for one device.
   * If multiple device will be queried for the window open status, a better approach would
   * be to get the device list once and then filter out the devices of interest.
   * @param ain
   */
  getWindowOpen: IExt<boolean> = async (ain) => {
    const device = await this.getDevice(ain);
    return device?.hkr?.windowopenactiv === '1';
  };

  /*
   * WLAN
   */

  // Parse guest WLAN form settings
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  parseGuestWlanHTML(html: string) {
    throw new Error('Not Implemented');
    /*
    $ = cheerio.load(html);
    const form = $('form');
    const settings = {};

    $('input', form).each(function (i, elem) {
      let val;
      const name = $(elem).attr('name');
      if (!name) return;

      switch ($(elem).attr('type')) {
        case 'checkbox':
          val = $(elem).attr('checked') == 'checked';
          break;
        default:
          val = $(elem).val();
      }
      settings[name] = val;
    });

    $('select option[selected=selected]', form).each(function (i, elem) {
      const val = $(elem).val();
      const name = $(elem).parent().attr('name');
      settings[name] = val;
    });

    return settings; */
  }

  // get guest WLAN settings - not part of Fritz API
  async getGuestWlan(): Promise<any> {
    const body = await this.executeCommand<string>(
      true,
      null,
      null,
      '/wlan/guest_access.lua?0=0',
    );
    return this.parseGuestWlanHTML(body);
  }

  // set guest WLAN settings - not part of Fritz API
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async setGuestWlan(enable: boolean) {
    throw new Error('Not Implemented');

    /*  /!* jshint laxbreak:true *!/
      const settings =
        enable ||
        executeCommand(
          sid,
          undefined,
          undefined,
          options,
          '/wlan/guest_access.lua?0=0'
        ).then(function (body) {
          return extend(parseGuestWlanHTML(body), {
            activate_guest_access: enable,
          });
        });

      return settings.then(function (settings) {
        // convert boolean to checkbox
        for (const property in settings) {
          if (settings[property] === true) settings[property] = 'on';
          else if (settings[property] === false) delete settings[property];
        }

        const req = {
          method: 'POST',
          form: extend(settings, {
            sid,
            xhr: 1,
            no_sidrenew: '',
            apply: '',
            oldpage: '/wlan/guest_access.lua',
          }),
        };

        return httpRequest('/data.lua', req, options).then(function (body) {
          return parseGuestWlanHTML(body);
        });
      }); */
  }

  /**
   * get phone devices
   */
  async getPhoneList() {
    return this.executeCommand(
      true,
      null,
      null,
      '/fon_num/foncalls_list.lua?csv=',
    );
  }
}
